pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,252 @@
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}).
+			Owns(&secv1.SecretProviderClass{}), manager.GetLogger(),
+	).Complete(&GatewaySecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (g *GatewaySecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, retErr error) {
+	// set up metrics given result/error
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(gatewaySecretProviderControllerName, res, retErr)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("creating logger: %w", err)
+	}
+	logger = gatewaySecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	// retrieve gateway resource from request + log the get attempt, but ignore not found
+	gwObj := &gatewayv1.Gateway{}
+	err = g.client.Get(ctx, req.NamespacedName, gwObj)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "failed to fetch Gateway")
+			return ctrl.Result{}, fmt.Errorf("fetching gateway: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if gwObj.Spec.GatewayClassName != istioGatewayClassName {
+		return ctrl.Result{}, nil
+	}
+
+	// check its TLS options - needs to have both cert uri and either serviceaccount name or clientid
+	for index, listener := range gwObj.Spec.Listeners {
+		spc := &secv1.SecretProviderClass{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "secrets-store.csi.x-k8s.io/v1",
+				Kind:       "SecretProviderClass",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      GenerateGwListenerCertName(gwObj.Name, listener.Name),
+				Namespace: req.Namespace,
+				Labels:    manifests.GetTopLevelLabels(),
+				OwnerReferences: []metav1.OwnerReference{{
+					APIVersion: gwObj.APIVersion,
+					Controller: util.ToPtr(true),
+					Kind:       gwObj.Kind,
+					Name:       gwObj.Name,
+					UID:        gwObj.UID,
+				}},
+			},
+		}
+		logger = logger.WithValues("spc", spc.Name)
+
+		if listenerIsKvEnabled(listener) {
+			var clientId string
+			clientId, err = retrieveClientIdFromListener(ctx, g.client, req.Namespace, listener.TLS.Options)
+			if err != nil {
+				var userErr userError
+				if errors.As(err, &userErr) {
+					logger.Info(fmt.Sprintf("failed to fetch clientId for SPC for listener %s due to user error: %s, sending warning event", listener.Name, userErr.userMessage))
+					g.events.Eventf(gwObj, corev1.EventTypeWarning, "InvalidInput", "invalid TLS configuration: %s", userErr.userMessage)
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, fmt.Sprintf("failed to fetch clientId for listener %s: %q", listener.Name, err.Error()))
```

Comment:
why %q?

Reply:
left over from when I was quoting errs but forgot to switch this back - updating!